### ISSUE TITLE
Corrupting the MDBX database

### ISSUE THREAD

**Issue by Eligioo** [nimiq/core-rs-albatross#1042]
## New issue checklist

- [x] I have read all of the [`README`](https://github.com/nimiq-network/core/blob/master/README.md) 
- [x] I have searched [existing issues](https://github.com/nimiq-network/core/issues?q=is%3Aissue+sort%3Acreated-desc) and **this is not a duplicate**.

### General information

As I was moving the MDBX database around on different servers the databases becomes corrupted as soon as you use it on the new machine even though it shouldn't store any env-specific information. The logs didn't show any useful information other than that the database is corrupted. I managed to enable logs on the underlying C library that the `libmdbx` crate is using to at least get a direction why this is happening. It has to with unusable `meta-pages` when it's loading the database. The location where the error is thrown is here: https://github.com/vorot93/libmdbx-rs/blob/master/mdbx-sys/libmdbx/mdbx.c on line 14557 within the `read_header` method. Also I have the logs that were created by the C library:

```
global_ctor:26705 pid 98137, &mdbx_rthc_key = 0x1036279c0, value 0x10a
 DEBUG nimiq_client         | Final configuration: ClientConfig { ... }
 INFO  nimiq_client         | Initializing client
 DEBUG file_store           | Reading from: /Users/stefan/.nimiq/peer_key.dat
 INFO  client               | Identity public key: Ed25519(PublicKey(compressed): af658aa03bee2a401128e67fa8b6a61ecf61e8d1d7df16911e317c1dbb0)
 INFO  client               | PeerId: 12D3KooWMd2sTVr9EJGB9yBxzceAAir81r9GoQyaRn98Hp1Xe1CT
 DEBUG client               | listen_addresses = ["/ip4/0.0.0.0/tcp/9100/ws"]
 DEBUG peer_contacts        | Adding peer contact peer_id=12D3KooWMd2sTVr9EJGB9yBxzceAAir81r9GoQyaRn98Hp1Xe1CT
 INFO  config               | Opening database: devalbatross-history-consensus
check_fstat:30914 lck-file is too short (0), exclusive-lock needed
uniq_check:5264 uniq-probe: unique
setup_lck:16201 lck-setup: with-lck exclusive
osal_mmap:28594 UNPOISON_MEMORY_REGION(0x103fbc000, 16384) at 28594
read_header:14496 reading meta[0]: offset 0, bytes 256, retry-left 42
validate_meta:14247 meta[0] took pagesize 4096
validate_meta:14272 checking meta0 = root 3/179, geo 3/180-181/268435456 +1 -1, txn_id 121, Weak
read_header:14549 latch meta[0]
read_header:14496 reading meta[1]: offset 4096, bytes 256, retry-left 42
validate_meta:14272 checking meta1 = root 4/181, geo 3/182-183/268435456 +1 -1, txn_id 122, Weak
read_header:14549 latch meta[1]
read_header:14496 reading meta[2]: offset 8192, bytes 256, retry-left 42
validate_meta:14272 checking meta2 = root 6/178, geo 3/179-179/268435456 +1 -1, txn_id 120, Weak
read_header:14496 reading meta[0]: offset 0, bytes 256, retry-left 42
validate_meta:14272 checking meta0 = root 3/179, geo 3/180-181/268435456 +1 -1, txn_id 121, Weak
read_header:14496 reading meta[1]: offset 4096, bytes 256, retry-left 42
validate_meta:14272 checking meta1 = root 4/181, geo 3/182-183/268435456 +1 -1, txn_id 122, Weak
read_header:14496 reading meta[2]: offset 8192, bytes 256, retry-left 42
validate_meta:14272 checking meta2 = root 6/178, geo 3/179-179/268435456 +1 -1, txn_id 120, Weak
read_header:14557 no usable meta-pages, database is corrupted
uniq_check:5264 uniq-probe: unique
osal_lck_destroy:31091 0x12bf118e0 got exclusive, drown locks
osal_munmap:28607 UNPOISON_MEMORY_REGION(0x103fbc000, 16384) at 28607
 ERROR logging              | MDBX error: MDBX_CORRUPTED: Database is corrupted
 ERROR logging              |   caused by
 ERROR logging              |     MDBX_CORRUPTED: Database is corrupted
 ERROR logging              |     MDBX_CORRUPTED: Database is corrupted
^C WARN  signal_handling      | Received Ctrl+C. Closing client
global_dtor:4980 >> pid 98137
global_dtor:4989 == thread 0x10403c580, rthc 0x0, pid 98137, self-status 0xffffffffffffffff, left 0
thread_key_delete:4861 key = 266
global_dtor:5070 << pid 98137
```

**Comment by sisou** [comment]
Can you share how to enable debug logs in the underlying C lib? I also have database corruption regularly after restarting my laptop, and would like to know more.

**Comment by Eligioo** [comment]
@sisou Here is how you can get started in the fastest way:

- Clone the Rust crate with version 0.1.7: https://github.com/vorot93/libmdbx-rs/commit/2ef372ae1a40d8610eea87c7223b8a12248af369
- Inside `core-rs-albatross` in `database\Cargo.toml`, change `libmdbx = "0.1.7"` to `libmdbx = { path = "path-to-the-libmdbx-cloned-repo"}`
- In the cloned libmdbx repo, within `mdbx-sys/libmdbx/mdbx.c`, do a search and replace with `if (mdbx_log_enabled` to `if (!mdbx_log_enabled`
- Recompile `core-rs-albatross` with `cargo build` and run the `nimiq-client`

This way is a bit hacky but otherwise you had to recompile with several env variables all the time and update the build script.

**Comment by sisou** [comment]
I think I have the same issue on my laptop after restart: `no usable meta-pages`:

```
2022-09-24T10:17:13.636969Z INFO  config               | Opening database: devalbatross-history-consensus
check_fstat:30580 lck-file is too short (0), exclusive-lock needed
uniq_check:5270 uniq-probe: unique
mdbx_setup_lck:16083 lck-setup: with-lck exclusive
mdbx_mmap:28283 UNPOISON_MEMORY_REGION(0x7fb0e61b6000, 4096) at 28283
mdbx_read_header:14398 reading meta[0]: offset 0, bytes 256, retry-left 42
mdbx_validate_meta:14147 meta[0] took pagesize 4096
mdbx_validate_meta:14167 checking meta0 = root 84754/4946810, geo 3/4946822-4946823/268435456 +1 -1, txn_id 22951, Weak
mdbx_read_header:14445 latch meta[0]
mdbx_read_header:14398 reading meta[1]: offset 4096, bytes 256, retry-left 42
mdbx_validate_meta:14167 checking meta1 = root 1060/4947180, geo 3/4947192-4947193/268435456 +1 -1, txn_id 22952, Weak
mdbx_read_header:14445 latch meta[1]
mdbx_read_header:14398 reading meta[2]: offset 8192, bytes 256, retry-left 42
mdbx_validate_meta:14167 checking meta2 = root 5159/4946481, geo 3/4946493-4946494/268435456 +1 -1, txn_id 22950, Weak
mdbx_read_header:14398 reading meta[0]: offset 0, bytes 256, retry-left 42
mdbx_validate_meta:14167 checking meta0 = root 84754/4946810, geo 3/4946822-4946823/268435456 +1 -1, txn_id 22951, Weak
mdbx_read_header:14398 reading meta[1]: offset 4096, bytes 256, retry-left 42
mdbx_validate_meta:14167 checking meta1 = root 1060/4947180, geo 3/4947192-4947193/268435456 +1 -1, txn_id 22952, Weak
mdbx_read_header:14398 reading meta[2]: offset 8192, bytes 256, retry-left 42
mdbx_validate_meta:14167 checking meta2 = root 5159/4946481, geo 3/4946493-4946494/268435456 +1 -1, txn_id 22950, Weak
mdbx_read_header:14453 no usable meta-pages, database is corrupted
uniq_check:5270 uniq-probe: unique
mdbx_lck_destroy:30758 0x563eac4856e0 got exclusive, drown locks
mdbx_munmap:28293 UNPOISON_MEMORY_REGION(0x7fb0e61b6000, 4096) at 28293
2022-09-24T10:17:13.638358Z ERROR logging              | MDBX error: MDBX_CORRUPTED: Database is corrupted
2022-09-24T10:17:13.638463Z ERROR logging              |   caused by
2022-09-24T10:17:13.638508Z ERROR logging              |     MDBX_CORRUPTED: Database is corrupted
2022-09-24T10:17:13.638550Z ERROR logging              |     MDBX_CORRUPTED: Database is corrupted
```

**Comment by sisou** [comment]
Using the `mdbx_chk` tool, I also cannot get it to check, as opening the DB fails with the same error. Forcing to use meta page `2`, I at least get an error that a specific leaf page is corrupted:
```
mdbx_chk v0.12.1-0-gb36a07a5 (2022-08-24T16:24:22+03:00, T-0803c79d2d94f2d1496166a9a86bd47da18c7eed)
Running for /home/soeren/.nimiq/devalbatross-history-consensus in 'read-write' mode...
Traversal b-tree by txn#22950...
 ! corrupted leaf-page #5159, mod-txnid 22953
 ! invalid page' txnid (22953) for parent-page' txnid (22950)
 - problems: invalid/corrupted (1)
Iterating DBIs...
Skip processing @MAIN since tree is corrupted (1 problems)
 ! abort processing '@GC' due to a previous error
 = Skipping turn to the specified meta-page (2) due to unsuccessful check!
Total 3 errors are detected, elapsed 0.035 seconds.
```

**Comment by sisou** [comment]
I also tried to dump/load the data from the db, following this guide (https://wiki.miranda-ng.org/index.php?title=Plugin:Dbx_mdbx), but even the dumper can't open the corrupted db, even when told to ignore errors.

**Comment by sisou** [comment]
I experimented again with the [`sync_mode` of the MDBX database](https://github.com/nimiq/core-rs-albatross/blob/albatross/database/src/mdbx.rs#L60), and found that the corruption on restart also happens for me with the `SafeNoSync` mode. But so far, with the `NoMetaSync` mode, it works correctly and a can re-use the database after restart.

**Comment by Eligioo** [comment]
@sisou is there any difference when using the NoMetaSync? A performance hit in some way?

**Comment by sisou** [comment]
Writes are supposed to be much slower with this setting, as per MDBX documentation. We tested it a while back in a same-machine closed validator setup, where it notably slowed the chain (when running without any block delays).

But since then we have added the 1s delay for empty blocks, plus in a real world-wide network blocks are propagated and produced slower anyway.

I'll try to dig out the theoretical maximum with that setting to see if it would fulfill our chain performance requirements.

As an alternative, we can also make it a config option. Remains to be seen if some slower validators would slow the whole chain.

**Comment by viquezclaudio** [comment]
We are going to restart our pubic devnet soon and we are planning to have 8 validators (let's call them main validators) running geographically distributed around the globe and we will have another 8 running with different sync mechanisms to test how they behave and test if they can keep up with the chain

**Comment by jsdanielh** [comment]
This issue was addressed in #1103.
